File: android/app/src/main/java/com/xiangshang/youth/core/model/DiagnosisReport.kt
package com.xiangshang.youth.core.model

enum class AssessmentRiskLevel(val label: String) { Low("正常关注"), Attention("建议关注"), High("重点跟进"), Unavailable("待完成/待复核") }

/** Central scoring policy for the seven school movement items. */
object AssessmentScoreRules {
    const val modelRegistryVersion = "UY-MODELS-1.0"
    const val algorithmVersion = "UY-IMCA-SCORE-1.3"
    const val calibrationVersion = "UY-CAL-BASELINE-1.0"
    const val itemCount = 7
    const val itemMaximum = 5.0
    const val totalMaximum = itemCount * itemMaximum
    const val lowItemThreshold = 3.0
    const val attentionTotalThreshold = 25.0
    const val highTotalThreshold = 21.0
    const val reviewConfidenceThreshold = .80
    const val duplicateConflictThreshold = 1.5

    fun score(value: Double): Double = if (!value.isFinite()) 0.0 else (kotlin.math.floor(value.coerceIn(0.0, itemMaximum) * 10.0 + 0.5) / 10.0)
    fun confidence(value: Double): Double = if (!value.isFinite()) 0.0 else value.coerceIn(0.0, 1.0)
    /** Match the backend's final aggregate half-up normalization. */
    fun total(value: Double): Double = if (!value.isFinite()) 0.0 else (kotlin.math.floor(value.coerceIn(0.0, totalMaximum) * 10.0 + 0.5) / 10.0)
}
data class DiagnosisReport(
    val id: String,
    val student: Student,
    val date: String,
    val scores: List<ScoreResult>,
    val abilityTags: List<String>,
    val riskAlerts: List<String>,
    val trainingAdvice: List<String>,
    val courseSuggestions: List<CourseSuggestion>,
    val ruleVersion: String,
    /** Regional policy snapshot used for this report's scoring explanation. */
    val regionPolicy: RegionPolicy,
    /** New reports carry audit metadata; old payloads decode via defaults. */
    val algorithmVersion: String = AssessmentScoreRules.algorithmVersion,
    val calibrationVersion: String = AssessmentScoreRules.calibrationVersion,
    val modelRegistryVersion: String = AssessmentScoreRules.modelRegistryVersion
) {
    companion object {
        /** Never show a bundled Mock report while the school service is loading. */
        fun unavailable(student: Student): DiagnosisReport = DiagnosisReport(
            id = "pending-${student.id}", student = student, date = "待学校服务返回", scores = emptyList(),
            abilityTags = emptyList(), riskAlerts = listOf("报告尚未从学校服务返回"),
            trainingAdvice = listOf("完成测评并等待学校发布报告"), courseSuggestions = emptyList(),
            ruleVersion = "pending", regionPolicy = RegionPolicy("pending", student.region, if (student.isPovertyArea) "重点帮扶地区" else null, "待返回", ""),
            algorithmVersion = "", calibrationVersion = "", modelRegistryVersion = ""
        )
    }
    /** Keep the strongest-evidence row when a malformed payload duplicates an item. */
    val normalizedScores: List<ScoreResult> get() = TestItem.entries.mapNotNull { item ->
        scores.filter { it.item == item }
            .maxWithOrNull(compareBy<ScoreResult> { AssessmentScoreRules.confidence(it.confidence) }
                .thenBy { if (it.reviewStatus == ScoreReviewStatus.Passed) 1 else 0 })
            ?.let { score ->
                val normalizedConfidence = AssessmentScoreRules.confidence(score.confidence)
                score.copy(
                    score = AssessmentScoreRules.score(score.score),
                    confidence = normalizedConfidence,
                    reviewStatus = if (!score.humanReviewed && normalizedConfidence < AssessmentScoreRules.reviewConfidenceThreshold) ScoreReviewStatus.PendingReview else score.reviewStatus
                )
            }
    }
    val scoredItemCount: Int get() = normalizedScores.map { it.item }.toSet().size
    val scoreCompletionRatio: Double get() = scoredItemCount.toDouble() / AssessmentScoreRules.itemCount
    val isComplete: Boolean get() = TestItem.entries.all { item -> normalizedScores.any { it.item == item } }
    val totalScore: Double get() = AssessmentScoreRules.total(normalizedScores.sumOf { it.score })
    val averageScore: Double get() = if (scoredItemCount == 0) 0.0 else totalScore / scoredItemCount
    val meanConfidence: Double get() = if (normalizedScores.isEmpty()) 0.0 else normalizedScores.map { it.confidence }.average()
    val lowScoreItems: List<ScoreResult> get() = normalizedScores.filter { it.score < AssessmentScoreRules.lowItemThreshold }
    val conflictingItems: List<TestItem> get() = TestItem.entries.filter { item ->
        val values = scores.filter { it.item == item }.map { AssessmentScoreRules.score(it.score) }
        val maximum = values.maxOrNull()
        val minimum = values.minOrNull()
        values.size > 1 && maximum != null && minimum != null &&
            (maximum - minimum) >= AssessmentScoreRules.duplicateConflictThreshold
    }
    val reviewItems: List<ScoreResult> get() = normalizedScores.filter { conflictingItems.contains(it.item) || it.reviewStatus == ScoreReviewStatus.PendingReview || (!it.humanReviewed && it.confidence < AssessmentScoreRules.reviewConfidenceThreshold) }
    val requiresReview: Boolean get() = !isComplete || reviewItems.isNotEmpty()
    val riskLevel: AssessmentRiskLevel get() = when {
        !isComplete || reviewItems.isNotEmpty() -> AssessmentRiskLevel.Unavailable
        totalScore < AssessmentScoreRules.highTotalThreshold || lowScoreItems.size >= 2 -> AssessmentRiskLevel.High
        totalScore < AssessmentScoreRules.attentionTotalThreshold || lowScoreItems.isNotEmpty() -> AssessmentRiskLevel.Attention
        else -> AssessmentRiskLevel.Low
    }
    val requiresFollowUp: Boolean get() = riskLevel == AssessmentRiskLevel.Attention || riskLevel == AssessmentRiskLevel.High || requiresReview
}

File: ios/XiangshangYouth/Core/Models/DiagnosisReport.swift
import Foundation

enum AssessmentRiskLevel: String, Codable {
    case low, attention, high, unavailable

    var label: String {
        switch self {
        case .low: "正常关注"
        case .attention: "建议关注"
        case .high: "重点跟进"
        case .unavailable: "待完成/待复核"
        }
    }
}

/// Central scoring policy for the seven school movement items. Scores coming
/// from the field device are still authoritative, but the app clamps and
/// explains malformed or incomplete data consistently on both platforms.
enum AssessmentScoreRules {
    static let modelRegistryVersion = "UY-MODELS-1.0"
    static let algorithmVersion = "UY-IMCA-SCORE-1.3"
    static let calibrationVersion = "UY-CAL-BASELINE-1.0"
    static let itemCount = TestItem.allCases.count
    static let itemMaximum = 5.0
    static let totalMaximum = Double(itemCount) * itemMaximum
    static let lowItemThreshold = 3.0
    static let attentionTotalThreshold = 25.0
    static let highTotalThreshold = 21.0
    static let reviewConfidenceThreshold = 0.80
    static let duplicateConflictThreshold = 1.5

    static func score(_ value: Double) -> Double {
        guard value.isFinite else { return 0 }
        return min(itemMaximum, max(0, floor(value * 10 + 0.5) / 10))
    }

    /// The API rounds the aggregate with the same non-bankers half-up rule as
    /// each item. Without this final normalization, binary floating-point
    /// sums such as 20.999999999 can cross the 21-point risk boundary on one
    /// client but not another.
    static func total(_ value: Double) -> Double {
        guard value.isFinite else { return 0 }
        return min(totalMaximum, max(0, floor(value * 10 + 0.5) / 10))
    }

    static func confidence(_ value: Double) -> Double {
        guard value.isFinite else { return 0 }
        return min(1, max(0, value))
    }
}

struct DiagnosisReport: Identifiable, Codable {
    let id: String; let student: Student; let assessmentDate: String; let scores: [ScoreResult]
    let abilityTags: [String]; let riskAlerts: [String]; let trainingAdvice: [String]; let courseSuggestions: [CourseSuggestion]; let ruleVersion: String
    /// A report must carry the regional policy snapshot that was used when it
    /// was produced.  The backend will supply the authoritative record later;
    /// Mock keeps the same immutable contract for UI and audit testing.
    let regionPolicy: RegionPolicy
    /// Version metadata is optional for reports generated before the audit
    /// contract was introduced; newly generated reports always carry it.
    var algorithmVersion: String? = AssessmentScoreRules.algorithmVersion
    var calibrationVersion: String? = AssessmentScoreRules.calibrationVersion
    var modelRegistryVersion: String? = AssessmentScoreRules.modelRegistryVersion

    /// Remote repositories must never render a bundled Mock report while an
    /// authoritative school report is still loading. This typed unavailable
    /// value keeps cards/layouts renderable without inventing scores.
    static func unavailable(for student: Student) -> DiagnosisReport {
        DiagnosisReport(id: "pending-\(student.id)", student: student, assessmentDate: "待学校服务返回", scores: [], abilityTags: [], riskAlerts: ["报告尚未从学校服务返回"], trainingAdvice: ["完成测评并等待学校发布报告"], courseSuggestions: [], ruleVersion: "pending", regionPolicy: RegionPolicy(id: "pending", region: student.region, povertyAreaLabel: student.isPovertyArea ? "重点帮扶地区" : nil, standardVersion: "待返回", effectiveDate: ""), algorithmVersion: nil, calibrationVersion: nil, modelRegistryVersion: nil)
    }
    var normalizedScores: [ScoreResult] {
        // A malformed sync payload can contain the same item twice. Keep the
        // strongest-evidence row so duplicates cannot inflate the total.
        TestItem.allCases.compactMap { item in
            scores.filter { $0.item == item }.max { lhs, rhs in
                let lhsConfidence = AssessmentScoreRules.confidence(lhs.confidence)
                let rhsConfidence = AssessmentScoreRules.confidence(rhs.confidence)
                if lhsConfidence != rhsConfidence { return lhsConfidence < rhsConfidence }
                // When evidence is tied, prefer the explicitly passed row;
                // it is the least surprising deterministic tie-breaker.
                return (lhs.reviewStatus == .passed ? 1 : 0) < (rhs.reviewStatus == .passed ? 1 : 0)
            }
        }.map { score in
            let normalizedConfidence = AssessmentScoreRules.confidence(score.confidence)
            let normalizedStatus: ScoreReviewStatus = !score.humanReviewed && normalizedConfidence < AssessmentScoreRules.reviewConfidenceThreshold ? .pendingReview : score.reviewStatus
            return ScoreResult(id: score.id, item: score.item, score: AssessmentScoreRules.score(score.score), note: score.note, confidence: normalizedConfidence, reviewStatus: normalizedStatus, humanReviewed: score.humanReviewed)
        }
    }

    var scoredItemCount: Int { Set(normalizedScores.map(\.item)).count }
    var scoreCompletionRatio: Double { Double(scoredItemCount) / Double(AssessmentScoreRules.itemCount) }
    var isComplete: Bool { TestItem.allCases.allSatisfy { item in normalizedScores.contains { $0.item == item } } }
    var totalScore: Double { AssessmentScoreRules.total(normalizedScores.reduce(0) { $0 + $1.score }) }
    var averageScore: Double { normalizedScores.isEmpty ? 0 : totalScore / Double(scoredItemCount) }
    var meanConfidence: Double { normalizedScores.isEmpty ? 0 : normalizedScores.map(\.confidence).reduce(0, +) / Double(normalizedScores.count) }
    var lowScoreItems: [ScoreResult] { normalizedScores.filter { $0.score < AssessmentScoreRules.lowItemThreshold } }
    var conflictingItems: [TestItem] {
        TestItem.allCases.filter { item in
            let values = scores.filter { $0.item == item }.map { AssessmentScoreRules.score($0.score) }
            guard values.count > 1, let minValue = values.min(), let maxValue = values.max() else { return false }
            return maxValue - minValue >= AssessmentScoreRules.duplicateConflictThreshold
        }
    }
    var reviewItems: [ScoreResult] { normalizedScores.filter { conflictingItems.contains($0.item) || $0.reviewStatus == .pendingReview || (!$0.humanReviewed && $0.confidence < AssessmentScoreRules.reviewConfidenceThreshold) } }
    var requiresReview: Bool { !isComplete || !reviewItems.isEmpty }
    var riskLevel: AssessmentRiskLevel {
        // A complete set of rows is not enough for a family-facing risk
        // label. Any low-confidence/pending row must be reviewed first;
        // otherwise a “low” label would look like a verified conclusion.
        guard isComplete, reviewItems.isEmpty else { return .unavailable }
        if totalScore < AssessmentScoreRules.highTotalThreshold || lowScoreItems.count >= 2 { return .high }
        if totalScore < AssessmentScoreRules.attentionTotalThreshold || !lowScoreItems.isEmpty { return .attention }
        return .low
    }
    var requiresFollowUp: Bool { riskLevel == .attention || riskLevel == .high || requiresReview }
}
